cpp/include/raft/spatial/knn/detail/ann_ivf_flat.cuh

```diff
@@ -0,0 +1,465 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../ann_common.h"
+#include "ann_ivf_flat_kernel.cuh"
+#include "ann_kmeans_balanced.cuh"
+#include "ann_utils.cuh"
+#include "topk/radix_topk.cuh"
+#include "topk/warpsort_topk.cuh"
+
+#include <raft/common/nvtx.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/cuda_utils.cuh>
+#include <raft/cudart_utils.h>
+#include <raft/distance/distance.cuh>
+#include <raft/distance/distance_type.hpp>
+#include <raft/linalg/gemm.cuh>
+#include <raft/linalg/unary_op.cuh>
+#include <raft/spatial/knn/ann_common.h>
+#include <raft/stats/histogram.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+#include <rmm/mr/device/pool_memory_resource.hpp>
+
+namespace raft::spatial::knn::detail {
+
+template <typename T>
+class cuivflHandle {
+ public:
+  cuivflHandle(const handle_t& handle,
+               raft::distance::DistanceType metric_type,
+               const ivf_flat_params& params);
+
+  void cuivflBuildIndex(const T* dataset, uint32_t n_rows, uint32_t dim);
+
+  void cuivflSetSearchParameters(const uint32_t n_probes,
+                                 const uint32_t max_batch,
+                                 const uint32_t max_k);
+
+  void cuivflSearch(
+    const T* queries, uint32_t n_queries, uint32_t k, size_t* neighbors, float* distances);
+
+  void queryIVFFlatGridSize(const uint32_t n_probes, const uint32_t n_queries, const uint32_t k);
+  uint32_t getDim() { return index_.has_value() ? index_->dim() : 0; }
+
+ private:
+  const handle_t& handle_;
+  const rmm::cuda_stream_view stream_;
+  ivf_flat_params params_;
+
+  const raft::distance::DistanceType metric_type_;
+  bool greater_;
```

Comment:
Consider the following common usage pattern for FAISS:

1. user trains an ivfflat index
2. user stores index to disk
3. user loads index from disk
4. user performs a search with index

In this case, what would be the indeded lifecycles of cuivflHandle and `ivf_flat_index`? Are we assuming the user would store the handle to disk or the index? I believe we're saying the index, right? So that means the handle is meant to be a temporary object which can be constructed on the fly each time a train or a search is called but doesn't necessarily need to be reused. 

If we removed the methods from `cuivflHandle` and allowed the `greater_` member to be derived from the `metric_type`, the only remaining members which would not be inputs from the functions would be `grid_dim_x_` and `search_mem_res`. It appears they are both reused across multiple calls to `search`. Maybe we could move them to the index itself and re-populated each time the index is instantiated (e.g. by not serializing them to disk).

Reply:
`ivf_flat_index` is now supposed to be the persistent part of the things. Without doing changes to the `metric_type`, I think, `metric_type` should be a part of `ivf_flat_index`, because the clusters are built with respect to the chosen metric (hence it does not make sense to have different metrics during building and searching).

I believe, `greater_` parameter could simply be a part of the search arguments. Though, I need to check in the code, how difficult it would be to not use it during `cuivflSetSearchParameters`/`queryIVFFlatGridSize`.

The `grid_dim_x_` member can be queried before search and it remains the same across searches if the batch size (`n_queries`) stays the same. I can check, I'm not sure if I can get rid of this parameter, but without it I'd like the code more.

The last member, `search_mem_res` is the memory resource/pool that can be reused across searches. It's there purely for performance reasons - to make expensive cudaMalloc* functions invoked only during the first search (or to preallocate a large enough buffer in advance).

With this in mind, I suggest the following:
  1. Keep all serialized state in `ivf_flat_index`
  2. Make either of the two:
       - (A)  non-persistent `ivf_flat_handle`, that would contain build/load/store/search methods and keep `ivf_flat_index` as an `optional<const>` member along with `search_mem_res` and possibly `grid_dim_x_`.
       - (B)  non-persistent `ivf_flat_search_state` that would keep  `search_mem_res` and possibly `grid_dim_x_`, and would be passed to the global public `search` function as an argument (build/load/store would be global public functions as well)

Which one would you prefer?

Reply:
> Without doing changes to the metric_type, I think, metric_type should be a part of ivf_flat_index, because the clusters are built with respect to the chosen metric (hence it does not make sense to have different metrics during building and searching).

Yep, 100% agreed the metric type should be stored in the index. 


> Which one would you prefer?

Does option 2A come w/ the implication that the users would need to construct a handle w/ an index? It's an extra step the users would need to take, unless maybe from the API perspective, they are loading the `handle` from disk instead of the index? Not a show-stopper, but it does seem like a little more awkward flow.

What do you think about maybe a combination of options 1 and 2B? We could create an `ivf_flat_search_state` and add it as a (non-persistent) member of the `ivf_flat_index`?

Reply:
The search state doesn't feel to belong to the index. Since you are more inclined towards option B, I'd prefer to not add an extra structure for the search state at all. I removed the `grid_dim_x_` already, so the only state left is the rmm memory resource. I think, we can just pass it as an optional argument, the same way as I do it now with `radix_topk`.
Afterwards, we can repeat the same discussion with the outermost `knnIndex` class :)

Reply:
> Afterwards, we can repeat the same discussion with the outermost knnIndex class :)

My take on your modifications to the `knnIndex` class is that you were trying to minimize public API changes by reusing the existing structures. I'm generally in favor of this approach but the goal here is to remove FAISS as a dependency, and so that will alleviate the need for this `knnIndex` class. My suggestion is that we add the new functions using `mdspan` and not worry about reusing the existing functions. The refactoring to cuml will be minimal anyways.